src/zcash/MMR.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef ZC_MMR_H_
+#define ZC_MMR_H_
+
+#include <stdexcept>
+#include <unordered_map>
+#include <boost/foreach.hpp>
+
+#include "serialize.h"
+#include "streams.h"
+#include "uint256.h"
+#include "librustzcash.h"
+
+namespace libzcash {
+
+const int NODE_SERIALIZED_LENGTH = 171;
+const int ENTRY_SERIALIZED_LENGTH = 180;
+
+typedef std::array<unsigned char, NODE_SERIALIZED_LENGTH> HistoryNode;
+typedef std::array<unsigned char, ENTRY_SERIALIZED_LENGTH> HistoryEntry;
+
+typedef long HistoryIndex;
+
+class HistoryCache {
```

Comment:
This should probably track the epoch it is for (which can just use the network upgrade index).